adrs/0001-cli-design.md

```diff
@@ -0,0 +1,306 @@
+
+### Manage operations
+A separate entrypoint for management operations is provided as convenience.
+
+In the short term, this area is meant for internal usage for managing core data services.
+
+```commandline
+> griffon manage osidb status 
+> griffon manage corgi status
+```
+
+### Plugins
+
+We **MAY** develop an extensible set of plugins to external services as an 'incubation' area for 
+integrating with third party services. This gives the CLI a good 'swiss army knife' utility as
+well as help identify integrations with more query/processes leveraging 3rd party services.
+
+A few example plugins are provided, for example, an example plugin integrating with https://osv.dev:
+
+```commandline
+> griffon z_osv query-by-commit-hash --commit_hash 6879efc2c1596d11a6a6ad296f80063b558d5e0f
+
+```
+Plugins are dynamically 'registered' once they are copied into the griffon/commands/plugins folder
+allowing individual end users to develop their own customisations.
+
+## Autocompletion
+
+The python click module [provides](https://click.palletsprojects.com/en/8.1.x/shell-completion/) provides custom shell completion.
+
+We **MAY** enable live autocompletion of data such as CVE-ids, product stream names/ofuris making it easier to use
+griffon.
+
+This can be achieved by generation of python module containing static values (ex. griffon/autocomplete/product_stream.py)
+or enhancing data services to provide autocompletion endpoints.
+
+## Process flow
+
+A lightweight process flow is defined as follows:
+
+1. CLI validate options/input
+2. invoke operation
+3. transform and emit output
+
+## Enhance services
+
+Some considerations for enhancing service bindings:
+
+* 'bake in' concurrent service requests (using threads or asyncio)
+* component-registry bindings need to be deployed on github and published to PyPI
+* component-registry bindings components.retrieve_list needs to support **ofuri**
+* retrieve a batch set of components with a list of {uuid|nvr|purl}
+* exception handling in bindings
+* implement next() in both bindings
+
+In addition, OSIDB needs to implement a 2nd affect type for mapping a single purl 
+to a CVE. It is unclear if this 2nd affect type accepts an array of purls or if we desire
+a one to one mapping ... impl detail left to the project.
+
+## Logging and Exception handling
+
+Fine grained logging is enabled using --debug flag
+```commandline
+griffon --debug entities components list --name curl
+```
+
+Most expected exceptions revolve around issues with http requests to services.
+
+## Optimising service calls
+
+parallel http requests using thread/asyncio
+
+## Decision(s)
+
+Develop a CLI (provisionally codenamed 'griffon') specifically integrating
+with security related entities useful for answering 'day to day' questions as 
+well and help automate some manual processes.
+
+Develop open source on github, using github actions for CI, with releases pushed
+to PyPI.
```

Comment:
I am convinced that we should aim for a very easy adoption of the tool so anyone (not just ProdSec) can start using it pretty much immediately. Ideally if something like **pip install griffon** would do everything needed for the very basic use - read-only public data. Then maybe something like **griffon setup** would guide the user quickly through creating a config file - eg. Bugzilla API key will be needed for flaw operations for Bugzilla time being. Then the help you describe above would do the rest.

Reply:
yes, added GRIF-30